test/functional/testnvim.lua

```diff
@@ -835,19 +835,156 @@ function M.exec_capture(code)
+local function set_upvalues(f, upvalues)
+  local i = 1
+  while true do
+    local n = debug.getupvalue(f, i)
+    if not n then
+      break
+    end
+    if upvalues[n] then
+      debug.setupvalue(f, i, upvalues[n])
+    end
+    i = i + 1
+  end
+end
+
+--- @type fun(f: function): table<string,any>
+_G.__get_upvalues = nil
+
+--- @type fun(f: function, upvalues: table<string,any>)
+_G.__set_upvalues = nil
+
+--- @param bytecode string
+--- @param upvalues table<string,any>
+--- @param ... any[]
+--- @return any[] result
+--- @return table<string,any> upvalues
+local function exec_lua_handler(bytecode, upvalues, ...)
+  local f = assert(loadstring(bytecode))
+  _G.__set_upvalues(f, upvalues)
+  local ret = { f(...) } --- @type any[]
+  local new_upvalues = _G.__get_upvalues(f)
+
+  do -- Check return value types for better error messages
+    local invalid_types = {
+      ['thread'] = true,
+      ['function'] = true,
+      ['userdata'] = true,
+    }
+
+    for k, v in pairs(ret) do
+      if invalid_types[type(v)] then
+        error(
+          string.format(
+            "Return index %d with value '%s' of type '%s' cannot be serialized over RPC",
+            k,
+            tostring(v),
+            type(v)
+          )
+        )
+      end
+    end
+  end
+
+  return ret, new_upvalues
+end
+
+--- Execute Lua code in the wrapped Nvim session.
+---
+--- When `code` is passed as a function, it is converted into Lua byte code.
+---
+--- Direct upvalues are copied over, however upvalues contained
+--- within nested functions are not. Upvalues are also copied back when `code`
+--- finishes executing. See `:help lua-upvalue`.
+---
+--- Only types which can be serialized can be transferred over, e.g:
+--- `table`, `number`, `boolean`, `string`.
+---
+--- `code` runs with a different environment and thus will have a different global
+--- environment. See `:help lua-environments`.
+---
+--- Example:
+--- ```lua
+--- local upvalue1 = 'upvalue1'
+--- exec_lua(function(a, b, c)
+---   print(upvalue1, a, b, c)
+---   (function()
+---     print(upvalue2)
+---   end)()
+--- end, 'a', 'b', 'c'
+--- ```
+--- Prints:
+--- ```
+--- upvalue1 a b c
+--- nil
+--- ```
+---
+--- Not supported:
+--- ```lua
+--- local a = vim.uv.new_timer()
+--- exec_lua(function()
+---   print(a) -- Error: a is of type 'userdata' which cannot be serialized.
+--- end)
+--- ```
 --- @param code string|function
+--- @param ... any[]
 --- @return any
 function M.exec_lua(code, ...)
-  if type(code) == 'function' then
-    return M.api.nvim_exec_lua(
+  if type(code) == 'string' then
+    return M.api.nvim_exec_lua(code, { ... })
+  end
+
+  assert(session)
+
+  if not session.exec_lua_setup then
+    M.api.nvim_exec_lua(
       [[
-      local code = ...
-      return loadstring(code)(select(2, ...))
+      _G.__get_upvalues = loadstring((select(1,...)))
+      _G.__set_upvalues = loadstring((select(2,...)))
+      _G.__exec_lua_handler = loadstring((select(3,...)))
```

Comment:
nit: store everything in a single `_G.__test` object?

```suggestion
      _G.__test = {
        get_upvalues = loadstring((select(1,...))),
        set_upvalues = loadstring((select(2,...))),
        exec_lua_handler = loadstring((select(3,...))),
      }
```

Reply:
done